crates/extensions/tedge_mqtt_bridge/src/lib.rs

```diff
@@ -40,22 +42,325 @@ use tedge_api::main_device_health_topic;
+            let rule = BridgeRule::new("a/topic", "", "");
+            assert_eq!(rule.apply("a/topic"), Some("a/topic".into()))
+        }
+
+        #[test]
+        fn forwards_wildcard_topics() {
+            let rule = BridgeRule::new("a/#", "", "");
+            assert_eq!(rule.apply("a/topic"), Some("a/topic".into()));
+        }
+
+        #[test]
+        fn does_not_forward_non_matching_topics() {
+            let rule = BridgeRule::new("a/topic", "", "");
+            assert_eq!(rule.apply("different/topic"), None)
+        }
+
+        #[test]
+        fn removes_local_prefix() {
+            let rule = BridgeRule::new("topic", "a/", "");
+            assert_eq!(rule.apply("a/topic"), Some("topic".into()));
+        }
+
+        #[test]
+        fn prepends_remote_prefix() {
+            // TODO maybe debug warn if topic filter begins with prefix to remove
+            let rule = BridgeRule::new("topic", "a/", "b/");
+            assert_eq!(rule.apply("a/topic"), Some("b/topic".into()));
+        }
+
+        #[test]
+        fn does_not_clone_if_topic_is_unchanged() {
+            let rule = BridgeRule::new("a/topic", "", "");
+            assert!(matches!(rule.apply("a/topic"), Some(Cow::Borrowed(_))))
+        }
+
+        #[test]
+        fn does_not_clone_if_prefix_is_removed_but_not_added() {
+            let rule = BridgeRule::new("topic", "a/", "");
+            assert!(matches!(rule.apply("a/topic"), Some(Cow::Borrowed(_))))
+        }
+    }
+
+    mod prepend {
+        use super::*;
+
+        #[test]
+        fn applies_nonempty_prefix_to_start_of_value() {
+            assert_eq!(prepend("tested", "being "), "being tested");
+        }
+
+        #[test]
+        fn does_not_clone_if_prefix_is_empty() {
+            assert!(matches!(prepend("test", ""), Cow::Borrowed(_)));
+        }
+
+        #[test]
+        fn leaves_value_unchanged_if_prefix_is_empty() {
+            assert_eq!(prepend("test", ""), "test");
+        }
+    }
+
+    mod single_converter {
+        use super::*;
+        #[test]
+        fn applies_matching_topic() {
+            let converter = SingleConverter(vec![BridgeRule::new("topic", "a/", "b/")]);
+            assert_eq!(converter.convert_topic("a/topic"), "b/topic")
+        }
+
+        #[test]
+        fn applies_first_matching_topic_if_multiple_are_provided() {
+            let converter = SingleConverter(vec![
+                BridgeRule::new("topic", "a/", "b/"),
+                BridgeRule::new("#", "a/", "c/"),
+            ]);
+            assert_eq!(converter.convert_topic("a/topic"), "b/topic");
+        }
+
+        #[test]
+        fn does_not_apply_non_matching_topics() {
+            let converter = SingleConverter(vec![
+                BridgeRule::new("topic", "x/", "b/"),
+                BridgeRule::new("#", "a/", "c/"),
+            ]);
+            assert_eq!(converter.convert_topic("a/topic"), "c/topic");
+        }
+    }
+}
+
+struct SingleConverter(Vec<BridgeRule>);
```

Comment:
```suggestion
struct TopicConverter(Vec<BridgeRule>);
```
In terms of naming, this feels more like the topic converter, and the existing struct named `TopicConverter` looks more like a `BridgeConfig` itself.